src/sage/rings/polynomial/integer_valued_polynomials.py

```diff
@@ -0,0 +1,1125 @@
+        is given by the sum
+
+        .. MATH::
+
+            \sum_{k=0}^{n_1} (-1)^k \binom{n_1}{k}\binom{n_1+n_2-k}{n_1} B[n_1 + n_2 - k].
+
+        There is a conversion formula between the two bases
+
+        .. MATH::
+
+            \binom{x}{i} = \sum_{k=0}^{i} (-1)^{i-k} \binom{i}{k} \binom{x+k}{k}.
+
+
+        REFERENCES:
+
+        - :wikipedia:`Integer-valued polynomial`
+
+        INPUT:
+
+        - ``A`` -- a parent with realization
+
+        EXAMPLES::
+
+            sage: F = IntegerValuedPolynomialRing(QQ).S(); F
+            Integer-Valued Polynomial Ring over Rational Field
+            in the shifted basis
+
+            sage: F.gen()
+            S[1]
+
+            sage: S = IntegerValuedPolynomialRing(ZZ).S(); S
+            Integer-Valued Polynomial Ring over Integer Ring
+            in the shifted basis
+            sage: S.base_ring()
+            Integer Ring
+
+            sage: G = IntegerValuedPolynomialRing(S).S(); G
+            Integer-Valued Polynomial Ring over Integer-Valued Polynomial
+            Ring over Integer Ring in the shifted basis in the shifted basis
+            sage: G.base_ring()
+            Integer-Valued Polynomial Ring over Integer Ring
+            in the shifted basis
+
+        Integer-valued polynomial rings commute with their base ring::
+
+            sage: K = IntegerValuedPolynomialRing(QQ).S()
+            sage: a = K.gen()
+            sage: K.is_commutative()
+            True
+            sage: L = IntegerValuedPolynomialRing(K).S()
+            sage: c = L.gen()
+            sage: L.is_commutative()
+            True
+            sage: s = a * c^3; s
+            S[1]*S[1] + (-6*S[1])*S[2] + 6*S[1]*S[3]
+            sage: parent(s)
+            Integer-Valued Polynomial Ring over Integer-Valued Polynomial
+            Ring over Rational Field in the shifted basis in the shifted basis
+
+        Integer-valued polynomial rings are commutative::
+
+            sage: c^3 * a == c * a * c * c
+            True
+
+        We can also manipulate elements in the basis and
+        coerce elements from our base field::
+
+            sage: F = IntegerValuedPolynomialRing(QQ).S()
+            sage: S = F.basis()
+            sage: S[2] * S[3]
+            3*S[3] - 12*S[4] + 10*S[5]
+            sage: 1 - S[2] * S[2] / 2
+            S[0] - 1/2*S[2] + 3*S[3] - 3*S[4]
+        """
+        def __init__(self, A):
+            r"""
+            Initialize ``self``.
+
+            INPUT:
+
+            - ``A`` -- a parent with realization
+
+            EXAMPLES::
+
+                sage: F = IntegerValuedPolynomialRing(QQ).S(); F
+                Integer-Valued Polynomial Ring over Rational Field
+                in the shifted basis
+                sage: TestSuite(F).run()
+            """
+            CombinatorialFreeModule.__init__(self, A.base_ring(),
+                                             A.indices(),
```

Comment:
I would just make this a direct instance call. If you want to attach it to the abstract object, perhaps as an attribute?

Reply:
I do not understand, sorry

Reply:
What I mean is
```
            CombinatorialFreeModule.__init__(self, A.base_ring(),
                                             NonNegativeIntegers(),
```